### ISSUE TITLE
Unable to guess signature of empty list/dict

### ISSUE THREAD

**Issue by belsander** [seveas/python-networkmanager#17]
```python
ERROR:dbus.connection:Unable to set arguments ({u'802-3-ethernet': {u's390-options': {}}, u'connection': {u'id': u'ethernet-eth0', u'timestamp': 1420558638, u'type': u'802-3-ethernet', u'uuid': 'dc38e52a-89e9-4c8c-93af-ad3e2bdd98dd', u'interface-name': u'eth0'}, u'ipv4': {u'routes': [], u'addresses': [[dbus.UInt32(1653084332L), dbus.UInt32(24L), dbus.UInt32(25694380L)]], u'dns': [], u'method': u'manual'}, u'ipv6': {u'routes': [], u'addresses': [], u'dns': [], u'method': u'ignore'}}, dbus.ObjectPath('/org/freedesktop/NetworkManager/Devices/1'), '/') according to signature u'a{sa{sv}}oo': <type 'exceptions.ValueError'>: Unable to guess signature from an empty dict
['  File "/usr/lib/pymodules/python2.7/Mgs/NetworkInterface.py", line 105, in __activate_connection\n    active_conn = NetworkManager.NetworkManager.AddAndActivateConnection(conn_settings, self.nm_device.object_path, "/")\n', '  File "/usr/lib/python2.7/dist-packages/NetworkManager.py", line 115, in proxy_call\n    ret = func(*args, **kwargs)\n', '  File "/usr/lib/python2.7/dist-packages/dbus/proxies.py", line 145, in __call__\n    **keywords)\n', '  File "/usr/lib/python2.7/dist-packages/dbus/connection.py", line 641, in call_blocking\n    message.append(signature=signature, *args)\n']
Unable to guess signature from an empty dict
```

I'm trying to execute AddAndActiveConnection, but I can't pass input parameters that are/contain an empty list/dict, because of the a{s**v**} signature. I've written a small function to replace empty lists and dicts by their dbus variant, but the parsing of the input parameters within python-networkmanager undoes what I parsed.
This function can be added to this lib, or maybe somebody has a better idea?

**Comment by afreof** [comment]
At least for me the problem is solved by the following patch. May be it should be mainlined...

From 5e5eb6345494d3b07568e9064c9841b8c90e43d4 Mon Sep 17 00:00:00 2001
From: Adrian <adrian.freihofer@gmail.com>
Date: Tue, 6 Jan 2015 15:38:15 +0100
Subject: [PATCH] Ignore empty dict

Signed-off-by: Adrian <adrian.freihofer@gmail.com>
---
 NetworkManager.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/NetworkManager.py b/NetworkManager.py
index 7865702..84020b5 100644
--- a/NetworkManager.py
+++ b/NetworkManager.py
@@ -94,7 +94,7 @@ class NMDbusInterface(object):
                     return val
         if hasattr(val, '__iter__') and not isinstance(val, basestring):
             if hasattr(val, 'items'):
-                return dict([(x, self.wrap(y)) for x, y in val.items()])
+                return dict([(x, self.wrap(y)) for x, y in val.items() if not (isinstance(y, (list, dict)) and not y)])
             else:
                 return [self.wrap(x) for x in val]
         return val
-- 
1.9.1

**Comment by belsander** [comment]
Yes, solves the problem for me as well. Thanks. I vote for mainline!

**Comment by seveas** [comment]
I don't like that idea very much. Why not incorporate your function that properly wraps `{}` and `[]`? That way everybody benefits :)

**Comment by afreof** [comment]
I agree, it's not a perfect solution. That's the reason why I did not publish it until somebody asked for it.
However, the empty dict is generated by NetworkManager itself. The following flow seems to be quite common but ends up in an exception:
  * read configuration from NetworkManager
  * modify the configuration
  * write the configuration back to NetworkManager.

The problem is the "u's390-options': {}" object which we get from NetworkManager. This triggers an exception if it is written back to NetworkManager.
The question is: Where is the right place to filter for objects in configurations which are provided by NetworkManager but not understood by NetworkManager?

**Comment by jasonharper** [comment]
Has any progress been made on fixing this?  I'm trying to write a simple HTML-based connection editor: the basic operation is to call GetSettings() on a connection, modify the resulting object (sample shown below), then call Update() to save the modifications.
`{u'connection': {u'timestamp': 1463427315, u'type': u'gsm', u'id': u'Cell', u'uuid': u'eb0a5040-6263-4ec2-b5ba-d6a20bcdb3b0'}, u'gsm': {u'apn': u'broadband'}, u'ipv4': {u'routes': [], u'addresses': [], u'dns': [], u'method': u'auto'}, u'ipv6': {u'routes': [], u'addresses': [], u'dns': [], u'method': u'auto'}}`
This fails every time with the "unable to guess signature" error, presumably due to all those empty arrays.  The patch above doesn't help - I think it handles empty dicts only.

**Comment by belsander** [comment]
Well, I was just lazy and patched it for our local distribution.

**Comment by knro** [comment]
Same here, patched it locally.

@jasonharper I've been looking for web based frontend for Network Manager, how is your project progressing? is it on GitHub?

**Comment by seveas** [comment]
This is fixed in a slightly different way in version 1.2.
